### PR TITLE
Make cycle_basis() deterministic

### PR DESCRIPTION
**Pull request by EfremBraun**
Fixes issue https://github.com/networkx/networkx/issues/6653

### CODE REVIEW

**Comment by EfremBraun** [comment]
Fixes issue #6653

**Review by dschult — COMMENTED** [review]
Using a list instead of a set makes the `remove` operation take much longer because removing an item from a list is O(n) while removing from a set or dict is O(1).

I recommend using a dict instead of a list. Using the keys of a dict is actually a decent way to store sets. And now that Python dicts are ordered, they give consistent results too.
```python
    gnodes = dict.fromkeys(G)
```
and then 
```python
        for node in pred:
            gnodes.pop(n)
```
Finally, can you come up with a short simple test that would test whether this works? It's good to test it so we make sure to keep the deterministic property after future changes.   

Thanks for this!

**Comment by EfremBraun** [comment]
Good ideas on both fronts. I'll implement tomorrow and post.

File: networkx/algorithms/cycles.py
Comment (EfremBraun) [reply networkx/algorithms/cycles.py]: Fixed!

**Comment by EfremBraun** [comment]
> Finally, can you come up with a short simple test that would test whether this works? It's good to test it so we make sure to keep the deterministic property after future changes.

I tried to do this, but unfortunately I'm only able to get the indeterminism on proprietary code that I can't share. I'm not sure exactly why it's showing indeterminism when my attempt to make a minimal reproducible example isn't. I copy/pasted the definition of the custom Node class, so I wouldn't thought that would do it. I'd probably have to spend quite a bit more time figuring out the exact reason for this; maybe a certain degree of complexity is required in the Graph that I'm not getting with my MRE.

Anyway, the code base I'm working with is HUGE, and figuring this out will take me a long time. The fix made in pull request does remove the indeterminism from my code, so I can personally say this fix works. Hopefully we're okay going without it, though obviously it would be better to have a regression test in place.

**Review by dschult — APPROVED** [review]
Good -- as I think about it more, I'm not sure how to test that order matters. Maybe we can find a simple case where the order does matter and run it with both orders. But that doesn't need to be part of this PR. I'll work on that separately.

Thanks very much!  I approve this PR

**Review by rossbar — APPROVED** [review]
LGTM - thanks @EfremBraun . Perhaps this fix should also get a release note, WDYT @dschult ?

**Comment by dschult** [comment]
Yes -- I think a release note -- and I'm close to a simple example for this too.  Can we hold off on merging for a day?
I'll remove my approval and approve it again when I've got that sorted out.

**Comment by dschult** [comment]
I found a test for the deterministic behavior: a graph with 2 cycles that when you change the names of two nodes gives a different (though equally valid) result.  

And I added a comment to the release_dev.  And, as usual it seems, I ask whether this should be a deprecation.   :)

**Comment by EfremBraun** [comment]
Looks great to me. Thanks @dschult!

**Review by dschult — APPROVED** [review]
Thanks for fixing the style typo.

I approve this PR.  @rossbar is this good to go?

**Comment by rossbar** [comment]
Looks great! Thanks @EfremBraun and @dschult !

**Pull request merged**
